src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/TopLevelCommandManager.cs

```diff
@@ -53,6 +54,88 @@ private async Task LoadTopLevelCommandsFromProvider(CommandProviderWrapper comma
         {
             TopLevelCommands.Add(new(new(i), true));
         }
+
+        commandProvider.CommandsChanged += CommandProvider_CommandsChanged;
+    }
+
+    private void CommandProvider_CommandsChanged(CommandProviderWrapper sender, ItemsChangedEventArgs args)
+    {
+        // By all accounts, we're already on a background thread (the COM call
+        // to handle the event shouldn't be on the main thread.). But just to
+        // be sure we don't block the caller, hop off this thread
+        _ = Task.Run(async () => await UpdateCommandsForProvider(sender, args));
+    }
+
+    /// <summary>
+    /// Called when a command provider raises its ItemsChanged event. We'll
+    /// remove the old commands from the top-level list and try to put the new
+    /// ones in the same place in the list.
+    /// </summary>
+    /// <param name="sender">The provider who's commands changed</param>
+    /// <param name="args">the ItemsChangedEvent the provider raised</param>
+    /// <returns>an awaitable task</returns>
+    private async Task UpdateCommandsForProvider(CommandProviderWrapper sender, ItemsChangedEventArgs args)
```

Comment:
Will the arguments be used? I noticed that on the bookmark extension, 0 was passed as an argument even though there were more than zero commands being changed/updated.

Reply:
You know, I haven't made up my mind on that point yet. 

I suspect there's some sort of future optimization we can do to like, insert "loading" items into the list between us getting the `ItemsChanged` and us actually retrieving the new items. 

That is, of course, hypothetical, so I keep just passing like, `0` or `-1` in a bunch of different places, since it ends up ignored for now. 🤷